pkg/kubectl/cmd/set/set_serviceaccount.go

```diff
@@ -0,0 +1,207 @@
+	output                 string
+	changeCause            string
+	local                  bool
+	categoryExpander       resource.CategoryExpander
+	clientMapper           resource.ClientMapper
+	typer                  runtime.ObjectTyper
+	namespace              string
+	enforceNamespace       bool
+	args                   []string
+	_print                 func(obj runtime.Object) error
+	updatePodSpecForObject func(runtime.Object, func(*api.PodSpec) error) (bool, error)
+}
+
+// NewCmdServiceAccount returns the "set service-account" command.
+func NewCmdServiceAccount(f cmdutil.Factory, out, err io.Writer) *cobra.Command {
+	saConfig := &serviceAccountConfig{
+		out: out,
+		err: err,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "service-account (-f FILENAME | TYPE NAME) SERVICE_ACCOUNT",
+		Aliases: []string{"sa"},
+		Short:   i18n.T("Update ServiceAccount of a resource"),
+		Long:    serviceaccountLong,
+		Example: serviceaccountExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(saConfig.Complete(f, cmd, args))
+			cmdutil.CheckErr(saConfig.Validate())
+			cmdutil.CheckErr(saConfig.Run())
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+
+	usage := "identifying the resource to get from a server."
+	cmdutil.AddFilenameOptionFlags(cmd, &saConfig.fileNameOptions, usage)
+	cmd.Flags().BoolVar(&saConfig.all, "all", false, "Select all resources in the namespace of the specified resource types")
+	cmd.Flags().BoolVar(&saConfig.local, "local", false, "If true, set image will NOT contact api-server but run locally.")
+	cmdutil.AddRecordFlag(cmd)
+	cmdutil.AddDryRunFlag(cmd)
+	return cmd
+}
+
+// Complete configures serviceAccountConfig from command line args.
+func (saConfig *serviceAccountConfig) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
+	saConfig.mapper, saConfig.typer = f.Object()
+	saConfig.encoder = f.JSONEncoder()
+	saConfig.shortOutput = cmdutil.GetFlagString(cmd, "output") == "name"
+	saConfig.record = cmdutil.GetRecordFlag(cmd)
+	saConfig.changeCause = f.Command(cmd, false)
+	saConfig.local = cmdutil.GetFlagBool(cmd, "local")
+	saConfig.dryRun = cmdutil.GetDryRunFlag(cmd)
+	saConfig.output = cmdutil.GetFlagString(cmd, "output")
+	saConfig.args = args
+	saConfig.decoder = f.Decoder(true)
+	saConfig.updatePodSpecForObject = f.UpdatePodSpecForObject
+	saConfig.clientMapper = resource.ClientMapperFunc(f.ClientForMapping)
+	saConfig.categoryExpander = f.CategoryExpander()
+	saConfig._print = func(obj runtime.Object) error {
+		return f.PrintObject(cmd, saConfig.local, saConfig.mapper, obj, saConfig.out)
+	}
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	saConfig.namespace, saConfig.enforceNamespace = cmdNamespace, enforceNamespace
+	return nil
+}
+
+// Validate does basic validation on serviceAccountConfig
+func (saConfig *serviceAccountConfig) Validate() error {
+	if len(saConfig.args) == 0 {
+		return errors.New("serviceaccount is required")
+	}
```

Comment:
Validation logic has been moved to Complete function. So no need for this function.